src/features/device-types/device-types.ts

```diff
@@ -159,21 +154,27 @@ async function fetchDeviceTypesAndReschedule(): Promise<
 	Dictionary<DeviceTypeInfo>
 > {
 	try {
-		const promise = withRetries(fetchDeviceTypes).then(
-			async (deviceTypeInfo) => {
-				// when the promise gets resolved, cache it
-				deviceTypesCache = promise;
-				return deviceTypeInfo;
-			},
-		);
+		const promise = withRetries(fetchDeviceTypes).then((deviceTypeInfo) => {
+			deviceTypesCache = Promise.resolve(deviceTypeInfo);
+			return deviceTypeInfo;
+		});
 
 		// if the cache is still empty, use this promise so that
 		// we do not start a second set of requests to s3
 		// in case another api request comes before the first completes
 		if (!deviceTypesCache) {
 			deviceTypesCache = promise;
 		}
-		return await promise;
+
+		try {
+			return await promise;
```

Comment:
Based on `deviceTypesCache` not needing to be an actual promise you could remove the `.then` above and replace it with
```suggestion
			deviceTypesCache = await promise;
			return deviceTypesCache;
```
or
```suggestion
			return deviceTypesCache = await promise;
```
here

Reply:
@Page-  The reason we are assigning the promise to the `deviceTypesCache`, is so that we do not start more than one concurrent `fetchDeviceTypes()` calls in case that multiple requests requiring this come in before we populate the cache.

Reply:
@thgreasi that's the reason for
```
		if (!deviceTypesCache) {
			deviceTypesCache = promise;
		}
```
above, to note I am *not* suggesting to remove that but instead to simplify how we replace the cached version on subsequent calls.

Reply:
Oh and you mean that we would also change:
```
let deviceTypesCache: Promise<Dictionary<DeviceTypeInfo>> | Dictionary<DeviceTypeInfo> | undefined;

// and then later just ...

		const promise = withRetries(fetchDeviceTypes);
		if (!deviceTypesCache) {
			deviceTypesCache = promise;
		}
		return (deviceTypesCache = await promise);
```
?
In that case I love your suggestion :heart: :heart: :heart:

Reply:
@Page- not sure that would work, if `deviceTypesCache` is not a promise, its value will only be set after `promise` resolves, which is too late to handle requests that come in the meantime.

Reply:
Let me add more :heart:s for this.
:heart: :heart: :heart: 
:heart: :heart: :heart:

Reply:
ohh I just read Thodoris' comment, if that is what you mean then yes, that would work :)

Reply:
Yeah, to expand further for clarity I'm thinking
```ts
async function fetchDeviceTypesAndReschedule(): Promise<
	Dictionary<DeviceTypeInfo>
> {
	try {
		const promise = withRetries(fetchDeviceTypes);

		// if the cache is still empty, use this promise so that
		// we do not start a second set of requests to s3
		// in case another api request comes before the first completes
		if (!deviceTypesCache) {
			deviceTypesCache = promise;
		}

		try {
			return deviceTypesCache = await promise;
		} catch (err) {
			if (deviceTypesCache === promise) {
				deviceTypesCache = undefined;
			}
			captureException(err, 'Failed to get device types');
			throw err;
		}
	} finally {
		// schedule a re-run to update the local cache - do not wait for it
		scheduleFetchDeviceTypes();
	}
}
```
and if you want the
```ts
		if (!deviceTypesCache) {
			deviceTypesCache = promise;
		}
```
could be replaced by
```ts
		deviceTypesCache ??= promise;

Reply:
Yeah ok that's how I got it after Thodoris' comment, I thought you are suggesting to not keep the deviceTypesCache as a promise.

Reply:
Yeah, more accurately I was suggesting it doesn't *always* need to be a promise